src/esp/gfx/PbrDrawable.cpp

```diff
@@ -142,6 +144,158 @@ void PbrDrawable::setMaterialValuesInternal(
   if (materialData_->isDoubleSided()) {
     flags_ |= PbrShader::Flag::DoubleSided;
   }
+  ////////////////
+  // ClearCoat layer
+  if (materialData_->hasLayer(Mn::Trade::MaterialLayer::ClearCoat)) {
+    const auto& ccLayer =
+        materialData_->as<Mn::Trade::PbrClearCoatMaterialData>();
+    float cc_LayerFactor = ccLayer.layerFactor();
+    if (cc_LayerFactor > 0.0f) {
+      // has non-trivial clearcoat layer
+      flags_ |= PbrShader::Flag::ClearCoatLayer;
+      //
+      matCache.cc_ClearCoatFactor = cc_LayerFactor;
+      matCache.cc_Roughness = ccLayer.roughness();
+
+      if (ccLayer.hasAttribute("layerFactorTexturePointer")) {
+        flags_ |= PbrShader::Flag::CCLayer_CCTexture;
+        matCache.cc_ClearCoatTexture =
+            ccLayer.attribute<Mn::GL::Texture2D*>("layerFactorTexturePointer");
+        matCache.cc_ClearCoatTexture_Swizzle =
+            ccLayer.layerFactorTextureSwizzle();
+      }
+
+      if (ccLayer.hasAttribute("roughnessTexturePointer")) {
+        flags_ |= PbrShader::Flag::CCLayer_RoughnessTexture;
+        matCache.cc_RoughnessTexture =
+            ccLayer.attribute<Mn::GL::Texture2D*>("roughnessTexturePointer");
+        matCache.cc_RoughnessTexture_Swizzle =
+            ccLayer.roughnessTextureSwizzle();
+      }
+      if (ccLayer.hasAttribute("normalTexturePointer")) {
+        flags_ |= PbrShader::Flag::CCLayer_NormalTexture;
+        matCache.cc_NormalTexture =
+            ccLayer.attribute<Mn::GL::Texture2D*>("normalTexturePointer");
+
+        matCache.cc_NormalTextureScale = ccLayer.normalTextureScale();
+        matCache.cc_NormalTexture_Swizzle = ccLayer.normalTextureSwizzle();
+      }
+
+    }  // non-zero layer factor
+  }    // has clearcoat layer
+
+  ////////////////
+  // KHR_materials_ior
+  if (materialData_->hasLayer("#KHR_materials_ior")) {
+    // Read in custom material index of refraction
+    if (materialData_->hasAttribute("#KHR_materials_ior", "ior")) {
+      matCache.ior_Index =
+          materialData_->attribute<float>("#KHR_materials_ior", "ior");
+    }
+
+  }  // has KHR_materials_ior layer
+
+  ////////////////
+  // KHR_materials_specular layer
+  if (materialData_->hasLayer("#KHR_materials_specular")) {
```

Comment:
While material attribute lookup by name is logarithmic, layer lookup by name is linear, so with each of those you have a `O(m + log n)`. I'll document that better, it doesn't seem to be stated clearly enough. There's [findLayerId()](https://doc.magnum.graphics/magnum/classMagnum_1_1Trade_1_1MaterialData.html#a1845070c3a94b31a749e318cb16bc887), use it like this:

```cpp
if(const Containers::Optional<UnsignedInt> specularLayerId = 
    materialData_->findLayerId("#KHR_materials_specular")) 
{
    materialData->attribute(specularLayerId, "...");
    ...
}    
```

And `hasAttribute()` and `attribute()` is also an unnecessary double lookup, so you can use the same pattern with `findAttributeId()` `findAttribute()`, `attributeOr()` etc. -- read the docs. It's also *significantly* less typing that way!

```cpp
if(const Containers::Optional<Float> specularFactor = 
    materialData_->findAttribute<Float>(specularLayerId, "specularFactor")) 
{
    matCache.spec_SpecularFactor = specularFactor;
}
```

This might apply also to the new code in https://github.com/facebookresearch/habitat-sim/pull/2083, I unfortunately didn't have the bandwidth to check everything there as well.